submariner/globalnet-enhancement2-0.md

```diff
@@ -0,0 +1,309 @@
+for both ingress and egress traffic can create Headless Services in front of them and should
+avoid requesting explicit globalIP that match Headless Service backend Pods.
+
+The proposal has its own set of Pros and Cons.
+
+### Pros
+
+Since we will be annotating only select K8s objects with globalIP, it will help in improved scalability.
+Also, this translates to fewer iptable rules on the active Gateway node of the cluster which will
+improve the overall performance of the solution.
+
+### Cons
+
+A limitation with the proposed approach is that we will lose the unique identity for **every**
+Pod which is supported today. For users that intend to use Network Policies across Clusters or
+for debugging purposes, this can be mitigated by creating the necessary GlobalnetEgressIP CRDs
+which create unique globalIPs at the Namespace/Pod level.
+
+### Caveats
+
+In the current implementation, when a client Pod is scheduled, until it's annotated with a globalIP,
+and the corresponding egress rules are programmed on the node, it was not able to communicate with
+remote Services. This was creating some issues in-terms of user-experience. In this proposal, as we
+have a globalIP at the cluster level, a Pod can communicate with remote Services right away which
+will be closer to the Vanilla Submariner experience.
+
+## Design Details
+
+The following CRD is proposed to support the new design.
+
+```Go
+type GlobalnetEgressIP struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Spec is the specification of desired behavior of GlobalnetEgressIP object.
+    Spec               EgressIPSpec `json:"spec"`
+
+    // Observed status of GlobalnetEgressIP. Its a read-only field.
+    // +optional
+    Status             EgressIPStatus `json:"status,omitempty"`
+}
+
+type EgressIPSpec struct {
+    // GlobalIPs is a list of EgressIPs requested. The requested globalIPs should be part
+    // of globalCIDR of the Cluster. If the requested globalIPs are not allocated to other
+    // K8s objects, then it will be allocated and the status field will be updated accordingly.
+    // If the globalIPs are already allocated to other objects, creation of the CRD will fail.
+    // If this field is not specified, a single globalIP will be allocated.
+    // This field cannot be changed through updates.
+    // +optional
+    GlobalIPs  []string `json:"globalIPs,omitempty"`
+
+    // Selects the Namespaces to which this GlobalnetEgressIP object applies.
+    // If an empty namespaceSelector: {} is configured, it selects all the Pods in all namespaces.
+    // On the other hand, if you omit specifying namespaceSelector, it does not select any
+    // namespaces and selects only Pods from the namespace where the GlobalnetEgressIP is
+    // deployed to.
+    // +optional
+    NamespaceSelector metav1.LabelSelector `json:"namespaceSelector,omitempty"`
```

Comment:
So in fact a CR with empty namespace and pod selectors is a cluster wide policy then?

Reply:
Good point. As commented in this PR, by default when Globalnet is deployed in a Cluster, it allocates a single (or configurable no of) globalIPs and there was no provision to request a specific globalIP at the cluster level. 

@mangelajo @mkolesnik I did not think about this configuration. Do you think we should allow this configuration and apply the Requested GlobalIPs (which is an optional field btw) cluster wide? or shall we simply return error?

Reply:
I'd create a default one with some reasonable default number of IPs (10?) when submariner is installed. We can also expose the pool size as an option for the installation, so that the users can specify it if they want.

I'd allow users to create a new one and delete the old one which is created by default. I don't see a specific problem with this. It's the same as allowing any other creation or deletion of resources.

If multiple "cluster" policies exist we shouldn't commit to which one takes precedence, in any case we could program multiple iptable rules and the first one will catch the packets, so it's really not a problem to have even multiple policies as long as users are informed that only one will be in effect at any given time.

Reply:
I like the idea of having the default one created. Then the cluster globalIPs are more explicit and can be consulted.

Like the "default" namespace.